File: 2025_08/11_Vowel_Balance/main_vowel_balance.py
# freecodecamp
# Daily Coding Challenges
# Day 1 (2025-08-11)
# Vowel Balance
"""
Given a string, determine whether the number of vowels in the first half of the string is equal to the number of vowels in the second half.
Vowels are: a, A, e, E, i, I, o, O, u, U
If there's an odd number of characters in the string, ignore the center character.
examples:
    1. is_balanced("racecar") should return True.
    2. is_balanced("Lorem Ipsum") should return True.
    3. is_balanced("Kitty Ipsum") should return False.
    4. is_balanced("string") should return False.
    5. is_balanced(" ") should return True.
    6. is_balanced("abcdefghijklmnopqrstuvwxyz") should return False.
    7. is_balanced("123A#b!E&*456-o.U") should return True.
"""


def count_vowels(string):
    vowels = ["a", "A", "e", "E", "i", "I", "o", "O", "u", "U"]
    count = 0

    for char in string:
        if char in vowels:
            count += 1

    return count



def is_balanced(s):
    half_count = int(len(s) / 2)
    string_one = s[0:half_count]
    
    if len(s) % 2 == 0:
        string_two = s[half_count:]
    else:
        string_two = s[half_count+1:]

    vowel_count_string_one = count_vowels(string_one)
    vowel_count_string_two = count_vowels(string_two)

    # for debugging --------------------------------------------
    # print(string_one)
    # print(string_two)
    # print(vowel_count_string_one)
    # print(vowel_count_string_one)
    
    return vowel_count_string_one == vowel_count_string_two

File: 2025_08/11_Vowel_Balance/main_vowel_balance.js
/*
freecodecamp
Daily Coding Challenges
Day 1 (2025-08-11)
Vowel Balance
*/

const vowelList = ["A", "a","E", "e","I", "i","O", "o","U", "u"]


function countVowels(text) {
    let len = text.length;
    let vowelLen = vowelList.length;
    let vowelCount = 0;
    
    for (let i=0; i < len; i++) {
        for (let j=0; j < vowelLen; j++) {
            if (text[i] === vowelList[j] ) {
                vowelCount++;
            }
        }
    }

    return vowelCount;
}


function getResult(count1, count2) {
    let result = false;
    
    if (count1 === count2) {
        result = true;
    }

    return result;
}

export function isBalanced(text) {
    let textLength = text.length;
    let halfTextLength = Math.floor(text.length / 2);
    
    let textPart1 = text.substr(0, halfTextLength);
    let textPart2 = "";
    
    if (textLength % 2 === 0) {
        textPart2 = text.substr(halfTextLength);
    }
    else {
        textPart2 = text.substr(halfTextLength + 1);
    }

    let vowelCount1 = countVowels(textPart1);
    let vowelCount2 = countVowels(textPart2);

    let result = getResult(vowelCount1, vowelCount2)

    return result;
}
